Embed html in a sandboxed iframe

mpkorstanje opened the pull request:
It preferable to allow the browser to render embedded  html documents. To ensure this is done safely the html is placed in a sandboxed iframe. The html is embeded in the srcdoc as recommend by  the w3c[1]. The src attribute is provided as a backup for legacy browsers.

  Additionally because the cucumber-reporting-plugin must necessarily serve embedded html documents from the same domain as the report  the download link will now force a popup to open. While this can not prevent the embedded content from tricking the user into opening it directly, it at least mitigates some of the issues mentioned by the w3c[1].

   1. https://www.w3.org/TR/html5/embedded-content-0.html#the-iframe-element

mpkorstanje commented:
Work in progress, open for discussion.

mpkorstanje commented on src/main/resources/templates/macros/json/embeddings.vm:
```diff
@@ -1,4 +1,4 @@
-#macro(includeEmbeddings, $embeddings)
+#macro(includeEmbeddings $embeddings)
```
The comma isn't needed here.

mpkorstanje commented on src/main/resources/templates/macros/json/embeddings.vm:
```diff
@@ -28,15 +28,33 @@
     <div data-toggle="collapse" data-target="#embedding-$embeddingId" class="collapsable-control">
       <a>Attachment $index ($image_type)</a>
-      <a href="embeddings/$embedding.getFileName()" target="_blank">
+      <a href="embeddings/$embedding.getFileName()" download target="_blank">
```
Ensures the download dialog is triggered

mpkorstanje commented on src/test/java/net/masterthought/cucumber/generators/AbstractPageTest.java:
```diff
@@ -64,7 +64,7 @@ public void generateReport_DisplaysContentAsEscapedText() {
                 .isEqualTo("As an Account Holder I want to withdraw cash from an ATM,<br>so that I can get money when the bank is closed");
         assertThat(document.getFeature().getElements()[0].getStepsSection().getSteps()[5].getEmbedding()[3].text())
-                .isEqualTo("Attachment 4 (HTML) <i>Hello</i> <b>World!</b>");
+                .isEqualTo("Attachment 4 (HTML)");
```
The element no longer contains any text. It is all contained in the srcDoc attribute of the iframe which we test in mbeddingAssertion.java

mpkorstanje commented:
@damianszczepanik  

can you fix sonar?

```ERROR] Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.2:sonar (default-cli) on project cucumber-reporting: You're not authorized to execute any SonarQube analysis. Please contact your SonarQube administrator. -> [Help 1]```

mpkorstanje commented on src/main/resources/css/cucumber.css:
```diff
@@ -77,6 +77,22 @@ a:hover {
+    width: 100%;
+    height: 100%;
+}
+
```
This ensures the iframe can show its full height

damianszczepanik commented:
Don't worry about Sonar, The problem is with permission for pull request and I have some idea how to make it working again

damianszczepanik commented:
So the way how it works on Jenkins depends of the security level you have and the content (if your file refer to external resources or not) ?

mpkorstanje commented:
> So the way how it works on Jenkins depends of the security level you have and the content (if your file refer to external resources or not) ?

mpkorstanje commented:
Erh. Github ate my reply while hitting the wrong button. I'll try answering later to day.

mpkorstanje commented:
We mostly want to protect the Jenkins session cookie. The session cookie is available to any JavaScript served from the same domain e.g. buildserver.local.

Embedding the HTML in a sandbox ensures that it can not access the cookie. However if the user visits the actual file at buildserver.local/path/to/embeddings/untrusted.html the cookie is available because it is served from the same domain. It is currently trivial to make that happen. The download link does it.

The common solution is to serve the untrusted content from a subdomain like usercontent.buildserver. local. However this is not possible with Jenkins.

The next best solution would be to ensure the mimetype for the untrusted.html is not text/html so the browser won't interpret it but I am not familiar with the Jenkins plugin.

mpkorstanje commented:
@damianszczepanik 

I've been looking at the SafeArchiveServingAction in cucumber-reports-plugin. Am I correct in assuming this bypasses the [CSP](https://wiki.jenkins-ci.org/display/JENKINS/Configuring+Content+Security+Policy#ConfiguringContentSecurityPolicy-TheDefaultRuleSet)?

damianszczepanik commented:
That's correct, make sure you haven't dismissed CSP by enabling the flag before Jnkins is started

mpkorstanje commented:
Mmh. The embeddings folder should be served with CSP. But we can treat that as a separate issue.

damianszczepanik commented:
It does but mind that HTML file may need external resources and then user may be surprised that page is incomplete.

damianszczepanik commented:
I have fixed Sonar - can you rebase the code so I can verify if it was really fixed? Sonar should not run any more for pull requests.

mpkorstanje commented:
Done.

mpkorstanje commented:
Fixed the tests and the styling. Good enough as far as I am concerned.

mpkorstanje commented:
Regarding the CSP. We can not trust the embedded documents so if they're not allowed to do anything that is just fine. The embedded documents can load their resources when they're in the sandbox iframe so that should be alright.

damianszczepanik commented:
Agree, my question was: do we want to inform user that the content may be incomplete? I had several issues related to CSP so I would rather not have several false positives issues that the plugin does not properly display content. What do you think?

mpkorstanje commented:
1. The CSP should not apply to for example path/to/reports/report-feature_net-masterthought-example-ATMK-feature.html. We can trust that file because we created it and took care to neuter the untrusted embedded content.

No warning should be needed here.

2. The CSP should apply to path/to/reports/embeddings/embedding_123456790.html because that link should only be used to download the file. Viewing it poses a security risk.

A warning would be nice here, but we have no where to show it.

3. I'm going to setup a Jenkins for testing to see how it works. Just to assure myself I understand it all correctly.

damianszczepanik commented:
One more thing to do: modify sample.json file and add sample HTML page with external resouce (eg Wikipedia logo) so it will be easier to detect regression. Consider add test for this as well.

mpkorstanje commented:
No problem. Bit short on time at the moment.

On Feb 20, 2017 23:24, "Damian Szczepanik" <notifications@github.com> wrote:

> One more thing to do: modify sample.json file and add sample HTML page
> with external resouce (eg Wikipedia logo) so it will be easier to detect
> regression. Consider add test for this as well.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/damianszczepanik/cucumber-reporting/pull/597#issuecomment-281196011>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGoAZ5XBWIFU4KDHR0Ev7TTgj-n4WL1Mks5rehKGgaJpZM4L0fCv>
> .
>

mpkorstanje commented:
I have added sample html to the existing hello world with a promt telling the viewer the wikipedia logo should appear. FeatureReportPageIntegrationTest.generatePage_generatesEmbedding covers this by using hasSrcDocContent to check if the proper content has been inserted in the srcDoc attribute of the iframe.

![image](https://cloud.githubusercontent.com/assets/6946919/23343084/137011ca-fc66-11e6-82ed-343f3c2b6970.png)

Rebased onto latest head.

mpkorstanje commented:
I have also established that cucumber-reports jenkins plugin is not properly applying the CSP for embeddings. I will file a separate issue for that in the appropriate repository. This does not block this fix though. The problem existed before this PR and will exist after it.

Unless you have further comments I think this PR ready.

mpkorstanje commented on src/main/resources/css/cucumber.css:
```diff
@@ -64,7 +64,7 @@ a:hover {
 .embedding-content {
     display: block;
-    padding: 9.5px;
+    padding: 10px;
```
No such thing as a half pixel.

mpkorstanje commented on src/test/java/net/masterthought/cucumber/generators/integrations/FeatureReportPageIntegrationTest.java:
```diff
@@ -306,7 +306,7 @@ public void generatePage_generatesEmbedding() {
         asserEmbeddingFileExist(embeddings[2]);
-        embeddingsElement[3].hasTextContent(embeddings[3].getData());
+        embeddingsElement[3].hasSrcDocContent(embeddings[3].getData());
         asserEmbeddingFileExist(embeddings[3]);
```
Checks if the srcDoc attribute contains the embedded html.

damianszczepanik commented:
Good job!

mpkorstanje commented:
You are welcome!
